src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPBreakpointSpanProvider.cs

```diff
@@ -0,0 +1,160 @@
+
+        private ILogger? _logHubLogger = null;
+
+        [ImportingConstructor]
+        public DefaultLSPBreakpointSpanProvider(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentSynchronizer documentSynchronizer,
+            RazorLogger razorLogger,
+            HTMLCSharpLanguageServerLogHubLoggerProvider loggerProvider)
+        {
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentSynchronizer is null)
+            {
+                throw new ArgumentNullException(nameof(documentSynchronizer));
+            }
+
+            if (razorLogger is null)
+            {
+                throw new ArgumentNullException(nameof(razorLogger));
+            }
+
+            if (loggerProvider is null)
+            {
+                throw new ArgumentNullException(nameof(loggerProvider));
+            }
+
+            _requestInvoker = requestInvoker;
+            _documentSynchronizer = documentSynchronizer;
+            _activityLogger = razorLogger;
+            _loggerProvider = loggerProvider;
+        }
+
+        public async override Task<Range?> GetBreakpointSpanAsync(LSPDocumentSnapshot documentSnapshot, Position position, CancellationToken cancellationToken)
+        {
+            if (documentSnapshot is null)
+            {
+                throw new ArgumentNullException(nameof(documentSnapshot));
+            }
+
+            if (position is null)
+            {
+                throw new ArgumentNullException(nameof(position));
+            }
+
+            // We initialize the logger here instead of the constructor as the breakpoint span provider is constructed
+            // *before* the language server. Thus, the log hub has yet to be initialized, thus we would be unable to
+            // create the logger at that time.
+            await InitializeLogHubAsync(cancellationToken).ConfigureAwait(false);
+
+            var languageQueryParams = new RazorBreakpointSpanParams()
+            {
+                Position = position,
+                Uri = documentSnapshot.Uri
+            };
+
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<RazorBreakpointSpanParams, RazorBreakpointSpanResponse>(
+                documentSnapshot.Snapshot.TextBuffer,
+                LanguageServerConstants.RazorBreakpointSpanEndpoint,
+                RazorLSPConstants.RazorLanguageServerName,
+                CheckRazorBreakpointSpanCapability,
+                languageQueryParams,
+                cancellationToken).ConfigureAwait(false);
+
+            var languageResponse = response?.Response;
+            if (languageResponse is null)
+            {
+                _logHubLogger.LogInformation("The breakpoint position could not be mapped to a valid range.");
+                return null;
+            }
+
+            VirtualDocumentSnapshot virtualDocument;
```

Comment:
Hmm, we probably don't need the rest of the code after this point. I thinkkkk we may be able to just return `languageResponse.Range` here. Reason being: Since we're asking a question to the server the server will indirectly ensure everything is synchronized / the document version is what we expect (via message order / document snapshot acquisition) etc.